src/cmd/partition.rs

```diff
@@ -0,0 +1,223 @@
+use std::collections::{HashMap, HashSet};
+use std::collections::hash_map::Entry;
+use std::fs;
+use std::io;
+use std::path::Path;
+
+use csv;
+use regex::Regex;
+use rustc_serialize::{Decodable, Decoder};
+
+use CliResult;
+use config::{Config, Delimiter};
+use select::SelectColumns;
+use util;
+
+static USAGE: &'static str = "
+Partitions the given CSV data into chunks based on the value of a column
+
+The files are written to the output directory with filenames based on the
+values in the partition column and the `--filename` flag.
+
+Usage:
+    xsv partition [options] <column> <outdir> [<input>]
+    xsv split --help
+
+partition options:
+    --filename <filename>  A filename template to use when constructing
+                           the names of the output files.  The string '{}'
+                           will be replaced by a value based on the value
+                           of the field, but sanitized for shell safety.
+                           [default: {}.csv]
+
+Common options:
+    -h, --help             Display this message
+    -o, --output <file>    Write output to <file> instead of stdout.
+    -n, --no-headers       When set, the first row will NOT be interpreted
+                           as column names. Otherwise, the first row will
+                           appear in all chunks as the header row.
+    -d, --delimiter <arg>  The field delimiter for reading CSV data.
+                           Must be a single character. (default: ,)
+";
+
+#[derive(Clone, RustcDecodable)]
+struct Args {
+    arg_column: SelectColumns,
+    arg_input: Option<String>,
+    arg_outdir: String,
+    flag_filename: FilenameTemplate,
+    // This option is unused, both here and in the split command.
+    flag_output: Option<String>,
+    flag_no_headers: bool,
+    flag_delimiter: Option<Delimiter>,
+}
+
+pub fn run(argv: &[&str]) -> CliResult<()> {
+    let args: Args = try!(util::get_args(USAGE, argv));
+    try!(fs::create_dir_all(&args.arg_outdir));
+
+    // It would be nice to support efficient parallel partitions, but doing
+    // do would involve more complicated inter-thread communication, with
+    // multiple readers and writers, and some way of passing buffers
+    // between them.
+    args.sequential_partition()
+}
+
+impl Args {
+    /// Configuration for our reader.
+    fn rconfig(&self) -> Config {
+        Config::new(&self.arg_input)
+            .delimiter(self.flag_delimiter)
+            .no_headers(self.flag_no_headers)
+            .select(self.arg_column.clone())
+    }
+
+    /// Get the column to use as a key.
+    fn key_column(&self, rconfig: &Config, headers: &[csv::ByteString])
+                  -> CliResult<usize> {
+        let select_cols = try!(rconfig.selection(&*headers));
+        if select_cols.len() == 1 {
+            Ok(select_cols[0])
+        } else {
+            Err("can only partition on one column".to_owned().into())
```

Comment:
Can you use the `fail!` macro in order to be consistent with the rest of the code? Thanks. Example: https://github.com/BurntSushi/xsv/blob/master/src/cmd/join.rs#L327